src/emitter/batch_emitter.rs

```diff
@@ -163,43 +180,75 @@ impl BatchEmitter {
 
             loop {
-                // `rx.recv().await` will not resolve until either a message is recieved,
+                // `rx.recv().await` will not resolve until either a message is received,
                 // or the channel is closed and there are no more messages, in which case we exit the loop
-                let message = match rx.recv().await {
-                    Some(msg) => {
-                        log::debug!("Received message: {:?}", msg);
-                        msg
-                    }
+
+                // select! is used to check both the `retry_rx` channel and the `rx` channel for new messages
+                let message = match tokio::select! {
+                    // `biased;` is used to ensure that the `retry_rx` channel is checked first, so retries get priority
+                    biased;
+
+                    retry = retry_rx.recv() => retry,
+                    event = rx.recv() => event,
+                } {
+                    Some(message) => message,
                     None => break,
                 };
 
                 match message {
-                    EmitterMessage::Send(batch) => {
-                        // Clone HttpClient to be moved into the task
+                    EmitterMessage::Send(mut batch) => {
+                        // Clone HttpClient and the retry transmitter to be moved into the task
                         let client = http_client.clone();
+                        let retry_transmitter = retry_tx.clone();
 
                         // Spawn a new task to send the batch
                         tokio_tasks.push(tokio::spawn(async move {
+                            if let Some(delay) = batch.delay {
+                                log::debug!("Delaying batch {} for {:?}", batch.id, delay);
+                                tokio::time::sleep(delay).await;
+
+                                if let Err(e) = batch.update_event_stm() {
+                                    // If the update fails, we just re-send the batch as-is
+                                    // Not ideal, but it's better than losing events
+                                    log::warn!(
+                                        "Failed to update stm of events in batch {} for retry: {e}",
+                                        batch.id
+                                    )
+                                };
+                            };
+
                             let batch_length = batch.events.len();
                             match Self::send_batch(batch, client).await {
-                                Ok(_) => {
-                                    log::info!("Sent batch of {batch_length} events")
+                                Ok(batch_id) => {
+                                    log::info!("Sent batch {batch_id} of {batch_length} events")
+                                }
+                                Err(mut failed_batch) => {
```

Comment:
Could we add a bit of logic here to check the status code of the HTTP response in order to decide whether to retry or not? We decided on a set of failure status codes for which sending events should not be retried (400, 401, 403, 410, 422). We have also started to make it possible for users to customize which status codes to retry on or not so that they can override this behaviour. The Java tracker implements the [logic for deciding whether to retry or not here](https://github.com/snowplow/snowplow-java-tracker/blob/master/src/main/java/com/snowplowanalytics/snowplow/tracker/emitter/BatchEmitter.java#L400-L414).